RFS/src/main/java/com/rfs/common/OpenSearchClient.java

```diff
@@ -243,27 +227,22 @@ public Optional<ObjectNode> getDocument(String indexName, String documentId) {
     }
 
     /*
-     * Update a document using optimistic locking.  Returns an Optional; if the document was updated, it
-     * will be the new value and empty otherwise.
+     * Update a document using optimistic locking with the versioning info in an original copy of the doc that is passed in.
+     * Returns an Optional; if the document was updated, it will be the new value and empty otherwise.
      */
-    public Optional<ObjectNode> updateDocument(String indexName, String documentId, ObjectNode body) {
-        Optional<ObjectNode> document = getDocument(indexName, documentId);
-        if (document.isEmpty()) {
-            throw new UpdateFailed("Document not found: " + indexName + "/" + documentId);
-        }
-
+    public Optional<ObjectNode> updateDocument(String indexName, String documentId, ObjectNode newBody, ObjectNode originalCopy) {
         String currentSeqNum;
         String currentPrimaryTerm;
         try {
-            currentSeqNum = document.get().get("_seq_no").asText();
-            currentPrimaryTerm = document.get().get("_primary_term").asText();
+            currentSeqNum = originalCopy.get("_seq_no").asText();
+            currentPrimaryTerm = originalCopy.get("_primary_term").asText();
```

Comment:
This field reference to which shard is the primary for this document.  Aren't documents only being processed once per shard.  It seems like if we are attempting to update documents on surprising shards something else has gone wrong and we shouldn't even attempt to make the request.

Reply:
I tried just supplying the sequence number, but the REST API balked at that.  Further investigation indicated that both fields are required to use the optimistic locking feature.